File: Level-1/42.java
class Solution {
    public int[] location(int loc){
        int[][] key = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}, {10, 0, 12}};
        
        int[] result = new int[2];
        
        for(int i = 0; i < key.length; i++){
            for(int j = 0; j < key[i].length; j++){
                if(key[i][j] == loc){
                    result[0] = i;
                    result[1] = j;
                }
            }
        }
        return result;
    }
    
    public String solution(int[] numbers, String hand) {
        String answer = "";
        int left = 10;
        int right = 12;
        
        for(int num : numbers){
            if(num == 1 || num == 4 || num == 7){
                left = num;
                answer += "L";
            }else if(num == 3 || num == 6 || num == 9){
                right = num;
                answer += "R";
            }else{
                int[] getNum = location(num);
                int[] getLeft = location(left);
                int[] getRight = location(right);
                
                int l = Math.abs(getLeft[0] - getNum[0]) + Math.abs(getLeft[1] - getNum[1]);
                int r = Math.abs(getRight[0] - getNum[0]) + Math.abs(getRight[1] - getNum[1]);
                
                if(l < r){
                    left = num;
                    answer += "L";
                }else if(l > r){
                    right = num;
                    answer += "R";
                }else{
                    if(hand.equals("left")){
                        left = num;
                        answer += "L";
                    }else if(hand.equals("right")){
                        right = num;
                        answer += "R";
                    }
                }
            }
        }
        return answer;
    }
}

File: Level-1/42.js
function solution(numbers, hand) {
    let answer = '';
    let left = '*';
    let right = '#';
    
    const findKey = (num) => {
        const numArr = [[1, 2, 3], [4, 5, 6], [7, 8, 9], ['*', 0, '#']];
        for(let i = 0; i < 4; i++) {
            for(let j = 0; j < 3; j++) {
                if(numArr[i][j] === num) return [i, j];
            }
        }
    };
    
    numbers.forEach((key) => {
        if(key === 1 || key === 4 || key === 7) {
            answer += 'L';
            left = key;
        }else if(key === 3 || key === 6 || key === 9) {
            answer += 'R';
            right = key;
        }else {
            const leftKey = findKey(left);
            const rightKey = findKey(right);
            const currKey = findKey(key);
            const leftDist = Math.abs(currKey[0] - leftKey[0]) + Math.abs(currKey[1] - leftKey[1]);
            const rightDist = Math.abs(currKey[0] - rightKey[0]) + Math.abs(currKey[1] - rightKey[1]);
            
            if(leftDist < rightDist) {
                answer += 'L';
                left = key;
            }else if(leftDist > rightDist) {
                answer += 'R';
                right = key;
            }else {
                if(hand === 'left') {
                    answer += 'L';
                    left = key;
                }else {
                    answer += 'R';
                    right = key;
                }
            }
        }
    });
    
    return answer;
}
